test/migrate.js

```diff
@@ -1,30 +1,45 @@
 /* exported expectWarning, expectNoWarning */
 
 function expectWarning( assert, name, expected, fn ) {
+	var result;
 	if ( !fn ) {
 		fn = expected;
 		expected = null;
 	}
 	jQuery.migrateReset();
-	fn();
+	result = fn();
 
-	// Special-case for 0 warnings expected
-	if ( expected === 0 ) {
-		assert.deepEqual( jQuery.migrateWarnings, [], name + ": did not warn" );
+	function check() {
 
-	// Simple numeric equality assertion for warnings matching an explicit count
-	} else if ( expected && jQuery.migrateWarnings.length === expected ) {
-		assert.equal( jQuery.migrateWarnings.length, expected, name + ": warned" );
+		// Special-case for 0 warnings expected
+		if ( expected === 0 ) {
+			assert.deepEqual( jQuery.migrateWarnings, [], name + ": did not warn" );
 
-	// Simple ok assertion when we saw at least one warning and weren't looking for an explict count
-	} else if ( !expected && jQuery.migrateWarnings.length ) {
-		assert.ok( true, name + ": warned" );
+		// Simple numeric equality assertion for warnings matching an explicit count
+		} else if ( expected && jQuery.migrateWarnings.length === expected ) {
+			assert.equal( jQuery.migrateWarnings.length, expected, name + ": warned" );
 
-	// Failure; use deepEqual to show the warnings that *were* generated and the expectation
-	} else {
-		assert.deepEqual( jQuery.migrateWarnings,
-			"<warnings: " + ( expected || "1+" ) + ">", name + ": warned"
+		// Simple ok assertion when we saw at least one warning and weren't looking for an explict count
+		} else if ( !expected && jQuery.migrateWarnings.length ) {
+			assert.ok( true, name + ": warned" );
+
+		// Failure; use deepEqual to show the warnings that *were* generated and the expectation
+		} else {
+			assert.deepEqual( jQuery.migrateWarnings,
+				"<warnings: " + ( expected || "1+" ) + ">", name + ": warned"
+			);
+		}
+	}
+
+	if ( result && result.then ) {
+		return jQuery.when(
```

Comment:
Note to self - `jQuery.when` is unavailable in jQuery 4.x Slim. Perhaps it's better to use a Promise polyfill like [native-promise-only](https://www.npmjs.com/package/native-promise-only) that we already use in Core tests.